cluster-autoscaler/cloudprovider/azure/azure_cache.go

```diff
@@ -20,97 +20,244 @@ import (
+func newAzureCache(client *azClient, cacheTTL time.Duration, resourceGroup, vmType string) (*azureCache, error) {
+	cache := &azureCache{
+		interrupt:            make(chan struct{}),
+		azClient:             client,
+		refreshInterval:      cacheTTL,
+		resourceGroup:        resourceGroup,
+		vmType:               vmType,
+		scaleSets:            make(map[string]compute.VirtualMachineScaleSet),
+		virtualMachines:      make(map[string][]compute.VirtualMachine),
+		registeredNodeGroups: make([]cloudprovider.NodeGroup, 0),
+		instanceToNodeGroup:  make(map[azureRef]cloudprovider.NodeGroup),
+		unownedInstances:     make(map[azureRef]bool),
 	}
 
-	cache.mutex.Lock()
-	defer cache.mutex.Unlock()
 	if err := cache.regenerate(); err != nil {
-		klog.Errorf("Error while regenerating Asg cache: %v", err)
+		klog.Errorf("Error while regenerating Azure cache: %v", err)
 	}
 
 	return cache, nil
 }
 
+func (m *azureCache) getVirtualMachines() map[string][]compute.VirtualMachine {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return m.virtualMachines
+}
+
+func (m *azureCache) getScaleSets() map[string]compute.VirtualMachineScaleSet {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	return m.scaleSets
+}
+
+// Cleanup closes the channel to signal the go routine to stop that is handling the cache
+func (m *azureCache) Cleanup() {
+	close(m.interrupt)
+}
+
+func (m *azureCache) regenerate() error {
+	err := m.fetchAzureResources()
+	if err != nil {
+		return err
+	}
+
+	// Regenerate instance to node groups mapping.
+	newInstanceToNodeGroupCache := make(map[azureRef]cloudprovider.NodeGroup)
+	for _, ng := range m.registeredNodeGroups {
+		klog.V(2).Infof("regenerate: finding nodes for node group %+v", ng)
+		instances, err := ng.Nodes()
+		if err != nil {
+			return err
+		}
+		klog.V(4).Infof("regenerate: found nodes for node group %v: %+v", ng, instances)
+
+		for _, instance := range instances {
+			ref := azureRef{Name: instance.Id}
+			newInstanceToNodeGroupCache[ref] = ng
+		}
+	}
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	m.instanceToNodeGroup = newInstanceToNodeGroupCache
+
+	// Reset unowned instances cache.
+	m.unownedInstances = make(map[azureRef]bool)
+
+	return nil
+}
+
+func (m *azureCache) fetchAzureResources() error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	switch m.vmType {
+	case vmTypeVMSS:
+		// List all VMSS in the RG.
+		vmssResult, err := m.fetchScaleSets()
+		if err == nil {
+			m.scaleSets = vmssResult
+		} else {
+			return err
+		}
+	case vmTypeStandard, vmTypeAKS:
+		// List all VMs in the RG.
+		vmResult, err := m.fetchVirtualMachines()
+		if err == nil {
+			m.virtualMachines = vmResult
+		} else {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// fetchVirtualMachines returns the updated list of virtual machines in the config resource group using the Azure API.
+func (m *azureCache) fetchVirtualMachines() (map[string][]compute.VirtualMachine, error) {
+	ctx, cancel := getContextWithCancel()
```

Comment:
`getContextWithTimeout` similar to how we were doing it before. With very large number of scale sets, I've seen the VMSS API take unrealistically long times for some of the GET calls and we don't want to hold the cache mutex lock for that long